arangod/Agency/TransactionBuilder.h

```diff
@@ -32,210 +32,214 @@
-    return write_trx(std::move(_buffer));
-  }
-
-  T done() && { _buffer.closeObject();
-    _buffer.setValue(VPackSlice::emptyObjectSlice());
-    _buffer.setValue(VPackValue(AgencyWriteTransaction::randomClientId()));
-    _buffer.closeArray();
-    return T(std::move(_buffer));
-  }
-
-  precs_trx<B, T> precs() && {
-    _buffer.closeObject();
-    _buffer.openObject();
-    return std::move(_buffer);
-  }
-  write_trx& operator=(write_trx&&) = default;
-
- private:
-  friend T;
-
-  explicit write_trx(buffer &&b) : _buffer(std::move(b)) {}
-  buffer _buffer;
-};
-
+namespace detail {
+template <typename T>
+struct moving_ptr {
+  moving_ptr(T* p) : _ptr(p) {}
+  ~moving_ptr() = default;
 
-template<typename B>
-struct envelope {
-  using buffer = buffer_mapper<B>;
+  moving_ptr& operator=(moving_ptr const&) = delete;
+  moving_ptr(moving_ptr const&) = delete;
 
-  read_trx<B, envelope> read() && {
-    _buffer.openArray();
-    return read_trx<B, envelope>(std::move(_buffer));
-  }
+  moving_ptr& operator=(moving_ptr&& o) {
+    reset(o.release());
+    return *this;
+  };
+  moving_ptr(moving_ptr&& o) { *this = std::move(o); }
 
-  write_trx<B, envelope> write() && {
-    _buffer.openArray();
-    _buffer.openObject();
-    return write_trx<B, envelope>(std::move(_buffer));
+  T* release() {
+    T* result = _ptr;
+    _ptr = nullptr;
+    return result;
   }
+  void reset(T* p) { _ptr = p; }
 
-  void done() && {
-    _buffer.closeArray();
-  }
-
-  static envelope create(VPackBuilder &b) {
-    buffer buff(b);
-    envelope env(buffer{b});
-    env._buffer.openArray();
-    return env;
-  }
+  T* operator->() { return _ptr; }
+  T* get() { return _ptr; }
+  T& operator*() { return *_ptr; }
+  operator bool() { return _ptr != nullptr; }
 
  private:
-  friend write_trx<B, envelope>;
-  friend read_trx<B, envelope>;
-  friend precs_trx<B, envelope>;
-  explicit envelope(buffer&& b) : _buffer(std::move(b)) {}
-  envelope() = default;
-  buffer _buffer;
+  T* _ptr = nullptr;
 };
 
-namespace detail {
-
-template<typename V>
+template <typename V>
 void add_to_builder(VPackBuilder* b, V const& v) {
   b->add(VPackValue(v));
 }
 
-template<>
+template <>
 inline void add_to_builder(VPackBuilder* b, VPackSlice const& v) {
   b->add(v);
 }
+}  // namespace detail
 
-}
-
-
-template<>
-struct buffer_mapper<VPackBuilder> {
-
-  explicit buffer_mapper(VPackBuilder& builder) : _builder(&builder) {};
-
-  template<typename K, typename V>
-  void setKey(K&& k, V&& v) {
-      setValue(std::forward<K>(k));
-      setValue(std::forward<V>(v));
+struct envelope {
+  using builder_ptr = detail::moving_ptr<VPackBuilder>;
+
+  struct read_trx {
+    envelope end() {
+      _builder->close();
+      return envelope(_builder.release());
+    }
+    template <typename K>
+    read_trx& key(K&& k) {
+      detail::add_to_builder(_builder.get(), std::forward<K>(k));
+      return *this;
+    }
+    ~read_trx() try {
+      if (_builder) {
+        end();
```

Comment:
I think calling `end()` here would work, but it is more work than necessary. `end()` closes the builder and reates a new envelope object. Why not just call `_builder.release()` here?